### PR TITLE
[SYCL] Enable handler copy method for N to M dimensions.

### PR DESCRIPTION
**Pull request by alexeyvoronov**
Enabled the handler copy method to use it for cases with
mismatched dimensions of accessors.

Signed-off-by: Alexey Voronov <alexey.voronov@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/handler.hpp [hunk 1]
```diff
@@ -1060,21 +1091,32 @@ class handler {
...
-#ifndef __SYCL_DEVICE_ONLY__
-    if (MIsHost) {
-      range<Dims_Src> Range = Dst.get_range();
+    // TODO replace to get_size() when it will provide correct values.
+    assert(
+        (Dst.get_range().size() * sizeof(T_Dst) >=
+         Src.get_range().size() * sizeof(T_Src)) &&
+        "dest must have at least as many bytes as the range accessed by src.");
+    if (MIsHost ||
+        !isCopyingBySliceAvailable(Src.get_range(), Dst.get_range())) {
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/handler.hpp]: Not sure I understand what isCopyingBySliceAvailable means and how a copy algorithm (will) differs depending on this check. Could you please clarify that?

Comment (alexeyvoronov) [reply sycl/include/CL/sycl/handler.hpp]: We can copy by slice(`clEnqueueCopyBufferRect`) under some conditions, for example:
case 2x2 to 3x3

```
12
34
```
->
```
000
000
000
```
 =
```
120
340
000
```

and in this case we can't (the source slice exceeds the destination slice)
case 4x2 to 2x4

```
1234
5678
```
->
```
00
00
00
00
```
=
```
1234
5678
00
00
```
but we can copy sequentially.
```
12
34
56
78
```
I plan to create an issue to describe all valid cases allowed for the copy operation and update CTS. The CTS do not cover all cases completely.

**Comment by keryell** [comment]
And what about https://gitlab.khronos.org/sycl/Specification/issues/72 ? Is there an impact ?

**Comment by alexeyvoronov** [comment]
> And what about https://gitlab.khronos.org/sycl/Specification/issues/72 ? Is there an impact ?

Yes,

```
>>> def sycl_id(id1,id2,id3,r1,r2,r3):
...     return id3 + (id2 * r3) + (id1 * r3 * r2)
...
>>> def opencl_id(id1,id2,id3,r1,r2,r3):
...     return id1 + (id2 * r1) + (id3 * r1 * r2)
...
>>> sycl_ids = []
>>> opencl_ids = []
>>> r1, r2, r3 = (2,2,2)
>>>
>>> for id1 in range(r1):
...     for id2 in range(r2):
...         for id3 in range(r3):
...             sycl_ids.append(sycl_id(id1,id2,id3,r1,r2,r3))
...             opencl_ids.append(opencl_id(id1,id2,id3,r1,r2,r3))
...
>>> print("SYCL", sycl_ids)
SYCL [0, 1, 2, 3, 4, 5, 6, 7]
>>> print("OpenCL", opencl_ids)
OpenCL [0, 4, 2, 6, 1, 5, 3, 7]
```

File: sycl/include/CL/sycl/handler.hpp [hunk 2]
```diff
@@ -1058,7 +1058,39 @@ class handler {
...
+  static id<3> getDelinearizedIndex(const range<3> Range, const size_t Index) {
+    size_t x = Index / (Range[1] * Range[2]);
+    size_t y = (Index / Range[2]) % Range[1];
+    size_t z = Index % Range[2];
+    return {x, y, z};
+  }
+
+  // Checks whether it is possible to copy the source shape to the destination
+  // shape(the shapes are described by the accessor ranges) by using
+  // piece-by-piece copying, where these shapes is a 1, 2 or 3 dimensional
```
[the remark this answers is not in the document]
Comment (alexeyvoronov) [reply sycl/include/CL/sycl/handler.hpp]: Done.

**Comment by keryell** [comment]
> >>> def sycl_id(id1,id2,id3,r1,r2,r3):
> ...     return id3 + (id2 * r3) + (id1 * r3 * r2)

Is this PyCL/PSYCL/PySYCL/...? :-)

**Comment by keryell** [comment]
The problem I see for know is what is the semantics for it?
There is no documentation describing it, even not a comment in the implementation...

**Comment by alexeyvoronov** [comment]
> > > > > def sycl_id(id1,id2,id3,r1,r2,r3):
> > > > > ...     return id3 + (id2 * r3) + (id1 * r3 * r2)
> 
> Is this PyCL/PSYCL/PySYCL/...? :-)

It's just Python, I copied the formulas from https://gitlab.khronos.org/sycl/Specification/issues/72 , and played with them a little.

**Comment by alexeyvoronov** [comment]
> The problem I see for know is what is the semantics for it?
> There is no documentation describing it, even not a comment in the implementation...

I made this patch to unlock the update for CTS https://gitlab.khronos.org/sycl/Specification/issues/181.

I agree that there are many pitfalls that should be described in the specification. Will we impose the same restrictions as OpenCL or not. For example, code snippet of copying below and other more complex options:
```
accessor<char, 1, access::mode::read, access::target::constant> AccessorSrc(BufferSrc, Cgh, range<1>(400));
accessor<int32_t, 2, access::mode::write, access::target::global_buffer> AccessorDst(BufferDst, Cgh, range<2>(10));
```

File: sycl/include/CL/sycl/handler.hpp [hunk 3]
```diff
@@ -1059,7 +1059,39 @@ class handler {
...
+    size_t z = Index % Range[2];
+    return {x, y, z};
+  }
+
+  // Checks whether it is possible to copy the source shape to the destination
+  // shape(the shapes are described by the accessor ranges) by using
+  // copying by regions of memory and not copying element by element
+  // Shapes can be 1, 2 or 3 dimensional rectangles.
+  template <int Dims_Src, int Dims_Dst>
+  static bool copyingRectRegionIsAvailable(const range<Dims_Src> Src,
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/handler.hpp]:
```suggestion
  static bool isCopyingByRegionAvailable(const range<Dims_Src> Src,
```

**Pull request merged**